accepted/single-file/design.md

```diff
@@ -66,225 +32,254 @@ Publishing to a single file can be triggered by adding the following property to
+    * In order to memory-map one embedded assembly at a time, the assemblies in the bundle must be aligned at 64KB boundaries.
+        * A prototype bundler with this behavior for testing purposes is available in this branch: https://github.com/swaroop-sridhar/core-setup/tree/single-exe
+
+We therefore map the assembly and perform an in-memory copy of the sections to appropriate offsets. In the long term, the solution to the mapping problem would involve considerations such as:
+* Compile all assemblies in a version bubble into one PE assembly, with a few aligned sections.
+* Embed the big composite assembly into the host with proper alignment, so that the single-exe bundle can be loaded without copies at run time.
+
+### API Semantics
 
-An app may choose to only embed some files (ex: due to licensing restrictions) and expect to pickup other dependencies from application-launch directory, nuget packages, etc. In order to resolve assemblies and native libraries, the embedded resources are probed first, followed by other probing paths. 
+#### `Assembly.Location`
 
-### New API
+There are a few options to consider for the `Assembly.Location` property of a bundled assembly:
 
-In this section, we propose adding a few APIs to facilitate common operations on bundled-apps.
+* A fixed literal (ex: `null`) indicating that no actual location is available.
+* The simple name of the assembly (with no path).
+* The path of the assembly as if it were not to be packaged into the single-file.
+* A special UNC notation such as `<bundle-path>/:/asm.dll` to denote files that come from the bundle. 
+* A configurable selection of the above, etc.
 
-The binaries that are published in the project are expected to be handled transparently by the host. However, explicit access to the embedded files is useful in situations such as:
+Proposed solution is: `:/<bundle-sub-path>/asm.dll`, where the `:` denotes the location within the bundle. The `:` is not a valid character in file-names on Windows. The `:` is already used as a path-separator on Unix systems.
 
-- Reading additional files packaged into the app (ex: data files).
-- Open an assembly for reflection/inspection
-- Load plugins built as single-file class-libs using existing Loader APIs.
+Most of the app development can be agnostic to whether the app is published as single-file or not. However, the parts of the app that deal with physical locations of files need to be aware of the single-file packaging. 
+
+#### `AppContext.BaseDirectory`
 
-Therefore, we propose adding an API similar to [GetManifestResourceStream](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.assembly.getmanifestresourcestream?view=netframework-4.7.2#System_Reflection_Assembly_GetManifestResourceStream_System_String_) to obtain a stream corresponding to an embedded file. 
+`AppContext.BaseDirectory` will be the directory where the AppHost (the single-file bundle itself) resides.
```

Comment:
Are we going to change this even for the self-extractor case, or is that one going to stay as is?

Reply:
There's currently no "self-extractor" option in .Net 5.
For netcoreapp3.x single file apps, the AppContext.BaseDirectory will still be the extraction directory, as noted in design_3_0.md [here](https://github.com/dotnet/designs/pull/90/files#diff-0472c74f7177d370f0ad8dbdbee2773aR93).